### ISSUE TITLE
npm notice created a lockfile as package-lock.json. You should commit this file.

### ISSUE THREAD

**Issue by samreid** [phetsims/chipper#578]
In npm 5.0.0, running `npm prune` in chipper gave this message:

```
~/github/chipper$ npm prune
npm notice created a lockfile as package-lock.json. You should commit this file.
up to date in 1.373s
```

Should we commit this file?  It seems so but let's learn more about it.

**Comment by samreid** [comment]
Documentation is here:
https://github.com/npm/npm/blob/latest/doc/files/package-lock.json.md

And good discussion is here:
https://www.reddit.com/r/javascript/comments/6dgnnq/npm_v500_released_save_by_default_lockfile_better/

I guess the next step is to have a team discussion about a strategy for moving (some or all team members) to NPM 5.

**Comment by samreid** [comment]
We discussed upgrading to latest node, which includes npm 5.0 on ox-dev, then we can test there.  @mattpen and @jbphet will work on it.

**Comment by mattpen** [comment]
I'll upgrade npm/node on ox-dev and test build-server, @jbphet will test rosetta.

**Comment by samreid** [comment]
@jonathanolson suggested making sure that all package-lock.jsons are identical across our repos.

**Comment by mattpen** [comment]
@jbphet - node 8/npm 5 are installed on phet-server-dev now.  Incidentally rosetta is checked out on the `mobile-app` branch.

**Comment by mattpen** [comment]
Currently builds are failing on ox-dev with this error:

```
Jun 01 2017 14:07:43 MDT - error: error running grunt checkout-master-all in ., build aborted to avoid infinite loop.
    /data/share/phet/phet-repos/perennial/node_modules/async/lib/async.js:30
    if (called) throw new Error("Callback was already called.");
    ^
    Error: Callback was already called.
        at /data/share/phet/phet-repos/perennial/node_modules/async/lib/async.js:30:31
        at /data/share/phet/phet-repos/perennial/js/build-server/build-server.js:342:11
        at ChildProcess.exithandler (child_process.js:259:5)
        at emitTwo (events.js:125:13)
        at ChildProcess.emit (events.js:213:7)
        at maybeClose (internal/child_process.js:887:16)
        at Socket.stream.socket.on (internal/child_process.js:340:11)
        at emitOne (events.js:115:13)
        at Socket.emit (events.js:210:7)
        at Pipe._handle.close [as _onclose] (net.js:546:12)
```

**Comment by samreid** [comment]
I don't think we can see the true error here.  Here is the context for the error:

```js
        if ( command === 'grunt checkout-master-all' ) {

          // checkout master for all repos if the build fails so they don't get left at random shas
          winston.log( 'error', 'error running grunt checkout-master-all in ' + dir + ', build aborted to avoid infinite loop.' );
          taskCallback( 'error running command ' + command + ': ' + err ); // build aborted, so take this build task off of the queue
        }
```

It would be good if we could see what the `err` was.

**Comment by mattpen** [comment]
The cause of the error was missing repos from active-repos.  Pulling master branch of chipper and running clone-missing-repos.sh fixed that problem.

However, builds are now failing due to grunt not being installed *locally*.  I've installed grunt and grunt-cli using the -g tags as sudo, but grunt commands do not work unless grunt was installed in a project using `npm install grunt`.  I don't seem to remember installing grunt in every directory as being a requirement for the build-server.

**Comment by samreid** [comment]
According to https://gruntjs.com/getting-started

```npm install -g grunt-cli``` This will put the grunt command in your system path, allowing it to be run from any directory.

The npm 5.0 docs said we also have to clean out all of the node_modules directories.  I did this on my machine by rm -rf ing them

**Comment by mattpen** [comment]
Thanks Sam.  For some reason, cleaning out the node_modules directories did it.  I used this one liner in the phet-repos directory:

`find . -name node_modules -exec rm -rf {} \;`

I had to commit package.json in chipper due to the new required trailing newline.  The build-server then ran successfully for an rc deploy, but failed during the cleanup tasks.  The error is below, but there isn't any details on what failed.  As the `rm -rf ./js/build-server/tmp` completed succesfully I'm guessing it failed either running `fs.mkdirSync( buildDir );` or during `writeDependenciesFile()`.

```
Jun 02 12:24:50 phet-server-dev.int.colorado.edu build-server[17537]: Jun 02 2017 12:24:50 MDT - info: rm -rf ./js/build-server/tmp ran successfully in directory: .
Jun 02 12:24:50 phet-server-dev.int.colorado.edu build-server[17537]: /data/share/phet/phet-repos/perennial/node_modules/async/lib/async.js:30
Jun 02 12:24:50 phet-server-dev.int.colorado.edu systemd[1]: build-server.service: main process exited, code=exited, status=1/FAILURE
Jun 02 12:24:50 phet-server-dev.int.colorado.edu systemd[1]: Unit build-server.service entered failed state.
Jun 02 12:24:50 phet-server-dev.int.colorado.edu systemd[1]: build-server.service failed.
Jun 02 12:24:50 phet-server-dev.int.colorado.edu systemd[1]: build-server.service holdoff time over, scheduling restart.
Jun 02 12:24:50 phet-server-dev.int.colorado.edu systemd[1]: Started build-server.service.
Jun 02 12:24:50 phet-server-dev.int.colorado.edu systemd[1]: Starting build-server.service...
```

**Comment by samreid** [comment]
Perhaps the cleanup failed because the npm 5.0 process created a new file package-lock.json which is not committed, so the working copy couldn't be restored?  Should we try committing the package-lock.jsons and see if that fixes it?

**Comment by mattpen** [comment]
A `deploy-production` for chains completed successfully with npm 5.0.

**Comment by mattpen** [comment]
> the working copy couldn't be restored

I don't think this is the problem because we haven't called `git add` yet, so it is effectively ignoring it for push and pull operations.  But it's possible and it probably wouldn't hurt.

**Comment by mattpen** [comment]
An intermittent error I've been seeing is:

```
Jun 02 12:54:33 phet-server-dev.int.colorado.edu build-server[21441]: Jun 02 2017 12:54:33 MDT - error: error running command: git checkout master in ../chipper, err: Error: Command failed: git checkout master
Jun 02 12:54:33 phet-server-dev.int.colorado.edu build-server[21441]: error: Your local changes to the following files would be overwritten by checkout:
Jun 02 12:54:33 phet-server-dev.int.colorado.edu build-server[21441]: package.json
Jun 02 12:54:33 phet-server-dev.int.colorado.edu build-server[21441]: Please, commit your changes or stash them before you can switch branches.
Jun 02 12:54:33 phet-server-dev.int.colorado.edu build-server[21441]: Aborting
Jun 02 12:54:33 phet-server-dev.int.colorado.edu build-server[21441]: , build aborted.
```

This seems to be temporarily fixed by calling `git stash; git checkout master` in chipper and restarting the build-server, but then pops up again on the second deploy.  I think we'll need to cherry-pick in the newline update to the package.json in chipper into the sha that chains is using to fix it permanently.

**Comment by samreid** [comment]
@mattpen and I discovered that:

```
npm config set save false
```

can be used to avoid having npm mess with whitespace at the end of package.jsons.  It looks like it also prevents creation of package-lock.json.

We learned about this here: https://github.com/npm/npm/issues/16823

**Comment by mattpen** [comment]
It can also be avoided temporarily by adding `--save=false` to `npm install` and `npm prune` calls.

**Comment by samreid** [comment]
We decided we would like @pixelzoom to weigh in before we start committing package-lock.json files or reformatted package.json files.  So for now, we should use npm 4 or npm 5 with --save=false (or config set save false) until we can take a closer look and discuss further.  

There is little problem upgrading master to have fixed package.jsons and package-lock.jsons--the issue is how to deal with this for legacy shas.  For instance, if wave on a string uses chipper sha X.  Then npm prune changes package.json in chipper sha X working copy, then we can no longer do git operations in chipper because it has a modified file on a sha (not a branch).

**Comment by mattpen** [comment]
To be discussed at dev meeting on 6/15.

**Comment by jbphet** [comment]
I tried testing rosetta, and ran into a few issues, see the commits shown above for details.  After making these changes, I was able to submit a test translation, but the build failed.  I can see the failure in the build server log, it looks like this:

```
Jun 02 16:38:38 phet-server-dev.int.colorado.edu build-server[21441]: error: Your local changes to the following files would be overwritten by checkout:
Jun 02 16:38:38 phet-server-dev.int.colorado.edu build-server[21441]: package.json
Jun 02 16:38:38 phet-server-dev.int.colorado.edu build-server[21441]: Please, commit your changes or stash them before you can switch branches.
Jun 02 16:38:38 phet-server-dev.int.colorado.edu build-server[21441]: Aborting
```
This sounds like it may be related to the issues that @mattpen ran into.  All other testing so far has succeeded, but until we have a successful end-to-end test I'm not comfortable saying that rosetta is ready for the upgrade.  I'll test again based on what is decided in the 6/15 dev meeting.

**Comment by pixelzoom** [comment]
I don't have time to do a deep dive into this discussion or the documentation.  Whatever the team decides is fine by me, as long as (a) it works with the build process, and (b) [sim-deployment.md](https://github.com/phetsims/phet-info/blob/master/deployment-info/sim_deployment.md) is updated to reflect any changes.

**Comment by samreid** [comment]
@jonathanolson suggests running a maintenance release process to add package-lock.json and updated package.json (added newline) to update all release branches.

**Comment by jonathanolson** [comment]
Something like example-server should have the lock file committed on a release branch. Then deploy-dev, deploy-rc, deploy-production and trigger a translation build to ensure those all work (on ox-dev).

Also make sure that the lock file, once created, would NOT be updated by "npm prune" and "npm update".

**Comment by samreid** [comment]
@ariel-phet recommended @jonathanolson and @mattpen to investigate further.

**Comment by mattpen** [comment]
I'll investigate and make sure that npm prune/update will not *change* the lock file once it has been created.

**Comment by mattpen** [comment]
As a future reference for when we upgrade node/npm on phet-server, this is what Mark and I did on phet-server-dev:

1. Run `grep node /etc/yum.repos.d/*` and then move the old repo definition for nodejs out to /tmp.
2. Run `curl --silent --location https://rpm.nodesource.com/setup_8.x | bash -` to setup yum repositories for Nodejs 8.
3. Run `yum clean all`
4. Run `yum install -y nodejs`
5. Verify the correct versions are installed with `node/npm -v`
6. Delete all the node_modules directories in the phetrepos directory
7. Restart build-server and test

**Comment by mattpen** [comment]
@samreid @jonathanolson - It looks like package-lock.json may change on its own.  According to [the docs](https://docs.npmjs.com/files/package-lock.json):

> package-lock.json is automatically generated for any operations where npm modifies either the node_modules tree, or package.json.

So both `prune` and `update` could potentially cause a re-generation of the lock file as they can both change the node_modules tree.  At least that's how I understand it, it appears there's been a lot of confusion, i.e. https://github.com/npm/npm/issues/16866

**Comment by mattpen** [comment]
We'll revisit this when we start work on revising the build tools.

**Comment by samreid** [comment]
@pixelzoom asked if we could just add package-lock.json to .gitignore

**Comment by pixelzoom** [comment]
If we add package-lock.json to .gitignore, then the file will still be in your local directory, and `npm update` will read it (which we don't want).

**Comment by ariel-phet** [comment]
Will investigate but not necessarily feeling like this is mandatory.

**Comment by mattpen** [comment]
I'm planning to upgrade node this week on phet-server and bayes, which will require a decision on this process.  In discussion for #604, we decided it was ok to upgrade if we make sure that the `--save=false` strategy is employed during npm operations (we can choose to change this later if needed).

**Comment by jonathanolson** [comment]
> I'm planning to upgrade node this week on phet-server and bayes

Should I shut down yotta/continuous-testing during this? Keep me apprised. @phet-steele, this may have continuous testing downtime?

> In discussion for #604, we decided it was ok to upgrade if we make sure that the --save=false strategy is employed during npm operations (we can choose to change this later if needed).

Does this require build-server/continuous-testing/automated-maintenance-release changes? Or do we need to set a global or per-user config?

**Comment by mattpen** [comment]
> Does this require build-server/continuous-testing/automated-maintenance-release changes? Or do we need to set a global or per-user config?

We can take either approach.  Using a user config file would probably be the best approach if we intend on using the package-lock files later.

> Should I shut down yotta/continuous-testing during this?

Yes and I definitely won't change anything without coordinating with you first.  The first step I'm planning to take is to make sure build-server, rosetta, and the phet-io learner-tokens work on phet-server-dev with the latest node.  After it's working there I'll upgrade it on phet-server and then move to bayes.  Hopefully the downtime will be minimal.

**Comment by mattpen** [comment]
Node and NPM have been updated on phet-server, phet-server-dev, and bayes.  We are able to start using package-lock.json now if we so choose, but it is currently disabled for the phet-admin user.  

To enable it, we should run `npm config set save true`

**Comment by jonathanolson** [comment]
This is in a "waiting for future decision if we use package-lock.json". Everything for current usage is complete.

**Comment by pixelzoom** [comment]
The upgrade to node8 and npm5 in https://github.com/phetsims/chipper/issues/604#issuecomment-345071159 turns this issue from "deferred" to "high priority".  I now have package-lock.json files in every repo where I've run `npm install`, and WebStorm shows all of them whenever I try to commit anything.

**Comment by samreid** [comment]
@pixelzoom did you use `npm config set save false` as described in https://github.com/phetsims/chipper/issues/578#issuecomment-305893183 and https://github.com/phetsims/chipper/issues/578#issuecomment-341777207 ?

**Comment by pixelzoom** [comment]
> @pixelzoom did you use `npm config set save false`

I see no definitive decision that we're doing that.  I see "waiting for future decision if we use package-lock.json", in  https://github.com/phetsims/chipper/issues/578#issuecomment-344690699

**Comment by samreid** [comment]
We are not using package-lock.json for now, in the future we will decide if we will use it in the future.

**Comment by pixelzoom** [comment]
I realize that we are not using package-lock.json now.  But #604 has now made a change where it is generated, when it previously not.  So the question is not whether to use it, it's how to ignore it.  If the decision is `npm config set save false`, great.  But I don't see any such decision in this issue.

**Comment by samreid** [comment]
I think we decided (but did not document) that each developer could choose whether to set the global or use `--save=false` in each call.  Much easier to set the global.

**Comment by pixelzoom** [comment]
Slack developer channel:

Chris Malley [3:58 PM] 
The most recent consensus about package-lock.json was “waiting for future decision if we use package-lock.json”, see https://github.com/phetsims/chipper/issues/578.  Are we going with `npm config set save false`, or something else?


Michael Kauzmann [4:01 PM] 
I set my global to false. Perhaps it would be best to do that, and then bring it up at a meeting soon?


Jonathan Olson [4:02 PM] 
Yes, I thought the decision was we all set it to false?


Chris Malley [4:02 PM] 
I see no such decision in the issue.


[4:03] 
I see a decision not to use package-lock.json, and to defer the issue until we started generating it.  I see no decision about how to ignore it.


Sam Reid [4:03 PM] 
I think we decided (but did not document) that each developer could choose whether to set the global or use `--save=false` in each call.  Much easier to set the global.


Chris Malley [4:03 PM] 
OK, let’s go with that. I’ll note in the issue.

**Comment by zepumph** [comment]
Should I add this to the developer meeting soon? Are we in the "future" and ready to commit these files? 

Decreasing priority now that things are documented better.

**Comment by mattpen** [comment]
> Are we in the "future" and ready to commit these files?

I investigated this in June.  Committing these would require doing a maintenance release of all or almost all of the sims in order for the build-server to work properly.  I thought we decided to punt because this is a steep cost for a small gain.

**Comment by zepumph** [comment]
Right! Thanks for the reminder.

**Comment by samreid** [comment]
> Committing these would require doing a maintenance release of all or almost all of the sims in order for the build-server to work properly.

Or adding logic to build server that can support old kinds + new kinds simultaneously?

**Comment by ankitakhangar** [comment]
npm notice created a lockfile as package-lock.json. You should commit this file.

**Comment by antorpramanik** [comment]
check file package-lock.json? (C:\Windows\system32)

if not exists
run cmd as admin
-------------------
Set EXPO_DEBUG=ture
npm config set package-lock false
npm install

now its working ;-)

**Comment by soniasingla** [comment]
Is this issue resolved? I am still facing this issue.

**Comment by samreid** [comment]
We decided to follow the approach prescribed in https://github.com/phetsims/chipper/issues/578#issuecomment-305893183.

Now that we have an established pattern for maintenance releases, I'll schedule for developer meeting discussion to see if we want to commit these files as discussed in https://github.com/phetsims/chipper/issues/578#issuecomment-345114259

**Comment by soniasingla** [comment]
@samreid that's great ! Thanks for the follow up 👍

**Comment by jonathanolson** [comment]
Running the npm command is in our documentation, and we'll want to decide on our method of pinning dependencies in https://github.com/phetsims/chipper/issues/114.

**Comment by samreid** [comment]
@jonathanolson observed that newer versions of node and npm do not respect `set save false`, but that `npm config set package-lock false` seems to work in those versions.
